druid/examples/event_viewer.rs

```diff
@@ -0,0 +1,466 @@
+// Copyright 2020 The Druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An application that accepts keyboard and mouse input, and displays
+//! information about received events.
+
+use druid::widget::{
+    prelude::*, Controller, CrossAxisAlignment, Flex, Label, List, Painter, Scroll, SizedBox,
+    TextBox,
+};
+use druid::{
+    theme, AppLauncher, Color, Data, Env, FontDescriptor, KeyEvent, Lens, LocalizedString,
+    Location, Modifiers, MouseButton, MouseEvent, Widget, WidgetExt, WindowDesc,
+};
+use std::sync::Arc;
+
+const WINDOW_TITLE: LocalizedString<AppState> = LocalizedString::new("Event Viewer");
+const INACTIVE_AREA_COLOR: Color = Color::grey8(0x55);
+const HOVER_AREA_COLOR: Color = Color::grey8(0xAA);
+const ACTIVE_AREA_COLOR: Color = Color::grey8(0xCC);
+const HEADER_BACKGROUND: Color = Color::grey8(0xCC);
+const COLUMN_PADDING: f64 = 2.0;
+const INTERACTIVE_AREA_DIM: f64 = 160.0;
+
+#[derive(Clone, Data, Lens)]
+struct AppState {
+    /// The text in the text field
+    text_input: String,
+    events: Arc<Vec<EventLog>>,
+    total_events: usize,
+}
+
+pub fn main() {
+    //describe the main window
+    let main_window = WindowDesc::new(build_root_widget)
+        .title(WINDOW_TITLE)
+        .window_size((760.0, 680.0));
+
+    //create the initial app state
+    let initial_state = AppState {
+        text_input: String::new(),
+        events: Arc::new(Vec::new()),
+        total_events: 0,
+    };
+
+    //start the application
+    AppLauncher::with_window(main_window)
+        .configure_env(|env, _| {
+            env.set(theme::UI_FONT, FontDescriptor::default().with_size(12.0));
+            env.set(theme::LABEL_COLOR, Color::grey8(0x11));
+        })
+        .launch(initial_state)
+        .expect("Failed to launch application");
+}
+
+fn build_root_widget() -> impl Widget<AppState> {
+    Flex::column()
+        .with_child(interactive_area())
+        .with_flex_child(event_list(), 1.0)
+}
+
+/// The top part of the application, that accepts keyboard and mouse input.
+fn interactive_area() -> impl Widget<AppState> {
+    Flex::row()
+        .with_flex_spacer(0.5)
+        .with_child(
+            TextBox::multiline()
+            .with_text_color(Color::rgb8(0xf0, 0xf0, 0xea))
+                .lens(AppState::text_input)
```

Comment:
huh, no clue :/

Reply:
looks like this gets screwed up by the `matches!` macro.

Reply:
It is because of the `matches!()` macros below this line, if you replace both the `matches!` calls in the statement with the output of `cargo expand` it formats fine.